libs-haskell/da-hs-base/tests/Telemetry.hs

```diff
@@ -0,0 +1,78 @@
+-- Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+module Telemetry(main) where
+
+import Control.Monad.Extra
+import DA.Daml.Project.Consts
+import DA.Service.Logger.Impl.GCP
+import DA.Service.Logger.Impl.Pure
+import DA.Test.Util
+import Data.Aeson (encode)
+import qualified Data.Text as T
+import System.Directory
+import System.Environment.Blank
+import System.IO.Extra
+import Test.Tasty
+import Test.Tasty.HUnit
+
+main :: IO ()
+main =
+  withTempDir $ \tmpDir1 -> do
+  withTempDir $ \tmpDir2 -> do
+  setPermissions tmpDir1 emptyPermissions
+  withEnv [("TASTY_NUM_THREADS", Just "1")] $
+    defaultMain $ do
+      testGroup "Telemetry" [dataLimiterTests, cacheDirTests tmpDir1 tmpDir2]
+
+dataLimiterTests :: TestTree
+dataLimiterTests =
+    testGroup
+        "data limiter"
+        [ withResource homeDir snd $ \_ ->
+              withResource
+                  (initialiseGcpState (GCPConfig "test" Nothing) makeNopHandle)
+                  (\gcpM -> whenJust gcpM $ \gcp -> removeFile (sentDataFile gcp)) $ \getGcp ->
+                  testCase "Check that limit is triggered" $ do
+                      gcpM <- getGcp
+                      case gcpM of
+                          Nothing -> fail "cache directory is not writable"
+                          Just gcp -> do
+                              let d = replicate 100 $ encode $ T.pack $ replicate 100000 'X'
+                              res <- mapM (sendData gcp fakeSend) d
+                              let l = length $ takeWhile isSuccess res
+                              83 @=? l
+        ]
+
+cacheDirTests :: FilePath -> FilePath -> TestTree
+cacheDirTests badCacheDir cacheDir =
+    testGroup
+        "cache directory"
+        [ withResource (setEnv damlCacheEnvVar cacheDir True) (\() -> unsetEnv damlCacheEnvVar) $ \_ ->
+              withResource
+                  (initialiseGcpState (GCPConfig "test" (Just badCacheDir)) makeNopHandle)
+                  (\gcpM -> whenJust gcpM $ \gcp -> removeFile (sentDataFile gcp)) $ \getGcp ->
+                  testCase
+                      "Check that opt-in/out, sendData, fetchMachineID works when a cache directory is specified." $ do
```

Comment:
A test that checks that a non-writable cache directory is supported as well would be great.

Reply:
I'll add one.